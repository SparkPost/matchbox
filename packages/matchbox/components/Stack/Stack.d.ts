export default Stack;
declare function Stack(props: any): any;
declare namespace Stack {
    const propTypes: {
        'data-id': any;
        /**
         * Sets the gutter space between children.
         * Styled-system responsive arrays work here.
         */
        space: any;
        /**
         * Positions children horizontally.
         * Accepts 'center', 'left', 'right', null.
         * Styled-system responsive arrays work here.
         */
        align: any;
    };
    const displayName: string;
}
