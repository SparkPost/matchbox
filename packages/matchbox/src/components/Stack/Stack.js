import React from 'react';
import { system } from 'styled-system';
import { Box } from '../Box';
import styled from 'styled-components';
import propTypes from '@styled-system/prop-types';

const StyledBox = styled(Box)`
  display: flex;
  flex-direction: column;
  ${() =>
    system({
      gutter: {
        property: 'paddingBottom',
        scale: 'space',
      },
    })}
  ${() =>
    system({
      alignment: {
        property: 'alignItems',
        defaultScale: {
          center: 'center',
          left: 'flex-start',
          right: 'flex-end',
        },
      },
    })}
`;

function Stack(props) {
  const { children, align, space = '500' } = props;
  const items = React.Children.toArray(children);

  if (items.length <= 1 && !align) {
    return <>{items}</>;
  }

  return (
    <div>
      {items.map((child, i) => (
        <StyledBox key={i} alignment={align} gutter={i < items.length - 1 ? space : null}>
          {child}
        </StyledBox>
      ))}
    </div>
  );
}

Stack.propTypes = {
  /**
   * Sets the gutter space between children.
   * Styled-system responsive arrays work here.
   */
  space: propTypes.space.paddingBottom,
  /**
   * Positions children horizontally.
   * Accepts 'center', 'left', 'right', null.
   * Styled-system responsive arrays work here.
   */
  align: propTypes.flexbox.alignItems,
};

Stack.displayName = 'Stack';
export default Stack;
