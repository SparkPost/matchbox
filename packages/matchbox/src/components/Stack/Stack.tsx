import React from 'react';
import { system } from 'styled-system';
import { Box } from '../Box';
import styled from 'styled-components';
import { ResponsiveValue } from 'styled-system';
import { AlignX } from '../../helpers/types';
import { SpaceKeys } from '../ThemeProvider/theme';

type AlignmentAndGutterProps = {
  $gutter?: ResponsiveValue<SpaceKeys | string>;
  $alignment?: ResponsiveValue<AlignX>;
};

const StyledBox = styled(Box)<AlignmentAndGutterProps>`
  display: flex;
  flex-direction: column;
  ${() =>
    system({
      $gutter: {
        property: 'paddingBottom',
        scale: 'space',
      },
    })}
  ${() =>
    system({
      $alignment: {
        property: 'alignItems',
        defaultScale: {
          center: 'center',
          left: 'flex-start',
          right: 'flex-end',
        },
      },
    })}
`;

type StackProps = {
  children?: React.ReactNode;
  /**
   * Positions children horizontally.
   * Styled-system responsive arrays work here.
   */
  align?: ResponsiveValue<AlignX>;
  /**
   * Sets the gutter space between children.
   * Styled-system responsive arrays work here.
   */
  space?: ResponsiveValue<SpaceKeys | string>;
  'data-id'?: string;
};

function Stack(props: StackProps): JSX.Element {
  const { children, align, space = '500' } = props;
  const items = React.Children.toArray(children);

  if (items.length <= 1 && !align) {
    return <>{items}</>;
  }

  return (
    <div data-id={props['data-id']}>
      {items.map((child, i) => (
        <StyledBox key={i} $alignment={align} $gutter={i < items.length - 1 ? space : null}>
          {child}
        </StyledBox>
      ))}
    </div>
  );
}

Stack.displayName = 'Stack';
export default Stack;
