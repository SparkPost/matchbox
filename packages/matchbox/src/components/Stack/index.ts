export { default as Stack } from './Stack';
export * from './Stack';
