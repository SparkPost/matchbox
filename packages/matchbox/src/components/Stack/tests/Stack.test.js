import React from 'react';
import Stack from '../Stack';
import 'jest-styled-components';

describe('Stack', () => {
  it('it should render defaults correctly', () => {
    const wrapper = global.renderStyled(<Stack><div>1</div><div>2</div></Stack>);
    expect(wrapper).toMatchSnapshot();
  });

  it('it should render with align and space props', () => {
    const wrapper = global.renderStyled(
      <Stack align='right' space='500'>
        <div>1</div><div>2</div>
      </Stack>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('it should render with align responsive props', () => {
    const wrapper = global.renderStyled(
      <Stack align={['center', 'left']}>
        <div>1</div>
      </Stack>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
