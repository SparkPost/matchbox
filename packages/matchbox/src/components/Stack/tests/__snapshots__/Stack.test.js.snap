// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack it should render defaults correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding-bottom: 400;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div>
  <div
    class="c0"
    display="flex"
  >
    <div>
      1
    </div>
  </div>
  <div
    class="c1"
    display="flex"
  >
    <div>
      2
    </div>
  </div>
</div>
`;

exports[`Stack it should render with align and space props 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  padding-bottom: 500;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

<div>
  <div
    class="c0"
    display="flex"
  >
    <div>
      1
    </div>
  </div>
  <div
    class="c1"
    display="flex"
  >
    <div>
      2
    </div>
  </div>
</div>
`;

exports[`Stack it should render with align responsive props 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

@media screen and (min-width:40em) {
  .c0 {
    -webkit-align-items: flex-start;
    -webkit-box-align: flex-start;
    -ms-flex-align: flex-start;
    align-items: flex-start;
  }
}

<div>
  <div
    class="c0"
    display="flex"
  >
    <div>
      1
    </div>
  </div>
</div>
`;
